basxconnect/mailer_integration/download_data.py

```diff
@@ -0,0 +1,69 @@
+from typing import NamedTuple
+
+from basxconnect.core import models
+from basxconnect.mailer_integration.abstract.abstract_datasource import Datasource
+from basxconnect.mailer_integration.abstract.abstract_mailer_person import MailerPerson
+from basxconnect.mailer_integration.models import Interest, MailingPreferences
+
+
+class SynchronizationResult(NamedTuple):
+    new_persons: int
+    total_synchronized_persons: int
+
+
+def download_persons(datasource: Datasource) -> SynchronizationResult:
+    mailer_persons = datasource.get_persons()
+    datasource_tag = _get_or_create_tag(datasource.tag())
+    new_persons = 0
+    for mailer_person in mailer_persons:
+        matching_persons = list(
+            models.Person.objects.filter(
+                primary_email_address__email=mailer_person.email(), deleted=False
+            ).all()
+        )
+        if len(matching_persons) == 0:
+            _save_person(datasource_tag, mailer_person)
+            new_persons += 1
+        else:
+            # if the downloaded email address already exists in our system, update the mailing preference of all persons
+            # that use this email address (as primary email address), without creating a new person in the database
+            for person in matching_persons:
+                _save_mailing_preferences(person, mailer_person)
+
+    return SynchronizationResult(new_persons, len(mailer_persons))
+
+
+def _is_new_person(
+    person: MailerPerson,
+) -> bool:
+    return not models.Person.objects.filter(
+        primary_email_address__email=person.email(), deleted=False
+    ).exists()
+
+
+def _get_or_create_tag(tag: str) -> models.Term:
+    tags_category_id = models.Category.objects.get(slug="category").id
+    return models.Term.objects.create(term=tag, category_id=tags_category_id)
```

Comment:
Should this not be get_or_create here?

Btw. here is another small "trick" for the Django ORM: You can assign database object directly to foreign keys, no need to specify the ID:

    models.Term.objects.create(term=tag, category=models.Category.objects.get(slug="category"))

Reply:
thanks!